var/spack/repos/builtin/packages/warpx/package.py

```diff
@@ -187,11 +187,16 @@ def cmake_args(self):
             self.define_from_variant("WarpX_QED_TABLE_GEN", "qedtablegen"),
         ]
 
+        # FindMPI needs an extra hint sometimes, particularly on cray systems
+        with when("+mpi"):
```

Comment:
Hi @kwryankrattiger,

I am sorry to be picky on this, but I need more information to accept this change please.

Regarding the problem:
- on which system did you see the problem
- what was your compiler config in Spack, versions of compiler and environment hints
- what was the CMake version
- what was the exact problem and for which `spec`? Can you post the failing CMake configure and/or failing build output?

You might be aware of our configurations on Cray systems, such as:
- Perlmutter: https://warpx.readthedocs.io/en/latest/install/hpc/perlmutter.html
- Crusher: https://warpx.readthedocs.io/en/latest/install/hpc/crusher.html
- Frontier: https://warpx.readthedocs.io/en/latest/install/hpc/frontier.html

Spack related question:
- Cray does not provide `mpicc`/`mpicxx` wrappers - does Spack add  them?

CMake uber question: I just want to make you aware my works in CMake, which should make these quirks unnecessary and help us to compile with with the Cray compiler wrappers as well as underlying vendor compilers
- CMake 3.22+: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6264
- WIP: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/7247

Reply:
@kwryankrattiger please don't be discouraged by my curiosity. I am truly interested in the underlying issue :)

Of course we will go forward with this to unblock our planned workflows :)

Reply:
Hey @ax3l No problem, I appreciate the feedback, makes things better long term :) 

To answer some of you questions:
> on which system did you see the problem

This was on Perlmutter, but I have found similar issues in Crusher testing unrelated to WarpX, but other projects that use MPI such as SENSEI and Ascent.

> what was your compiler config in Spack, versions of compiler and environment hints

I am using GCC 10.3.0, I didn't try with the CCE compiler, but previously I was having issues with it.

> what was the CMake version

CMake 3.23.2

> what was the exact problem and for which spec? Can you post the failing CMake configure and/or failing build output?

I am building `warpx+ascent+sensei` concretized unified with `ecp-data-vis-sdk+sensei+ascent`. I don't have the logs anymore, when I get a chance I will re-produce them and attach them here.

> Cray does not provide mpicc/mpicxx wrappers - does Spack add them?

So cray-mpich actually does provide these compiler wrappers, they are located in `$CRAY_MPICH_PREFIX/bin`. That said, they are not used to compiler anything, although they can be, only to assist CMakes `FindMPI` in locating the include and library directories. In an ideal world, CMake should be able to find these automagically, and when you configure a cmake project outside of spack it works just fine without the hints, but for some reason in the spack environment this additional hint is needed. I have tried digging a couple times to ascertain the reason why but I couldn't find anything in particular that would cause the issue.

>  let us limit it down, e.g., to platform cray.

The reason I didn't do this is because I have seen this kind of issue on non-cray systems, I have an AMD test bed that I work on for development that occasionally demonstrates this behavior. By putting this in general I figure it at very least can future proof against configurations that are weird for some unknown reason.

I suspect this is related to why SENSEI needs to define `SENSEI_DIR` even though it is in the `CMAKE_PREFIX_PATH` environment variable, which really doesn't make any sense. Whenever I get some time, I am going to give this a proper dig because it would be awesome to clean a lot of this stuff up.

> You might be aware of our configurations on Cray systems

I haven't seen this before. I will look through it more closely, but at a glance it seems like it is more or less mimicking what spack is/should be doing behind the scenes.

Reply:
Note that we would really like to move all Spack builds to use the `mpicc` wrappers with cray-mpich -- eventually `spec["mpi"].mpicc` will start returning those instead of `cc`, simply because there is so much garbage that `cc` puts on the compile line -- we can't control it effectively from within Spack.  This is part of moving away from modules and towards a build model for cray that looks much like every other platform.

So if you want to start using those *instead* of `cc` it would actually be great.  I would be interested to know if you run into issues.